Fix all clippy warnings and add clippy to the CI

koute opened the pull request:
Clippy doesn't like the theoretically redundant closure calls that I've added in my last PR, and since clippy doesn't run on the CI here I didn't notice.

While I was at it I also fixed all other clippy warnings, plus one error as it was complaining about an `unsafe` that I've now removed. I've also added clippy to the CI (hopefully it works).

Fixes https://github.com/paritytech/parity-scale-codec/issues/454

Fixes https://github.com/paritytech/parity-scale-codec/issues/456

ggwpez reviewed (commented):
Maybe we should add a Clippy config file?  
Could be the one from Substrate in `.cargo/config.toml`.

ggwpez commented on CHANGELOG.md:
```diff
@@ -5,6 +5,12 @@ All notable changes to this crate are documented in this file.
+
+### Fixed
+
+- Deriving `Decode` will not trigger `redundant_closure_call` clippy warnings anymore
```
And the unneeded return warning?

koute replied:
Yes, no warnings should appear now. (Or at least they don't appear now when I tested it running `cargo clippy` in `polkadot` :P)

ggwpez commented on src/codec.rs:
```diff
@@ -1093,7 +1094,7 @@ pub(crate) fn compact_encode_len_to<W: Output + ?Sized>(dest: &mut W, len: usize
 impl<T: Encode> Encode for [T] {
 	fn size_hint(&self) -> usize {
-		mem::size_of::<u32>() + mem::size_of::<T>() * self.len()
+		mem::size_of::<u32>() + core::mem::size_of_val(self)
```
Not sure if saturating add would make sense here.

koute replied:
Hmm... perhaps, but then it's just a size hint so shouldn't really matter if it overflows.

ggwpez replied:
I dont know if the runtime is normally compiled with overflow checks enabled (that would be my only concern).

bkchr replied:
Overflow checks are not enabled for release builds as they would be too slow.

ggwpez commented on src/compact.rs:
```diff
@@ -493,7 +493,7 @@ impl Decode for Compact<u32> {
 				}
 			},
-			3|_ => {	// |_. yeah, i know.
+			3 => {	// |_. yeah, i know.
```
```suggestion
			3 => {
```
:sparkles:

ggwpez commented on src/encode_append.rs:
```diff
@@ -80,25 +80,15 @@ impl<T: Encode> EncodeAppend for crate::alloc::collections::VecDeque<T> {
+/// Extends a SCALE-encoded vector with elements from the given `iter`.
+///
+/// `vec` must either be empty, or contain a valid SCALE-encoded `Vec<Item>` payload.
+fn extend_encoded_vec<Item, I>(
```
Nit: I think the work `append` is also used in the `sp-io` storage API to refer to this operation.

koute replied:
In `std` it's called `extend`, so that's why I called it like that. (:

But I don't particularly care. Do you want me to rename it?

koute replied:
Okay, I renamed it to be consistent with the method name in the trait where it's used.

ggwpez commented on src/encode_append.rs:
```diff
@@ -107,46 +97,41 @@ where
+			// We can't update the length as the new length prefix will take up more
+			// space when encoded, so we need to move our data to make space for it.
+			let mut new_vec = Vec::with_capacity(
+				new_item_count_encoded_bytesize + new_item_count as usize * mem::size_of::<Item>()
```
I dont quite understand the code. Why is it using `mem::size_of` instead of some `Encode::size_hint` or something?  
Or is there some assumed relation between in-memory and encoded size?

koute replied:
Ah, you're right! Good catch! It's quite late here so I didn't notice that and just copied what the code was originally doing here, but this *is* incorrect.

In fact, now this is harmless since it's only used for `with_capacity` (as I removed the `unsafe`, but previously with the `unsafe` it looks like we had a potential out-of-bounds write...  yikes.

koute replied:
I've simplified this. Not sure if there's a better way of doing this. (We can't really get the exact encoded bytesize of the new items since we haven't iterated over them yet.)

ggwpez reviewed (commented):
Looks good, just some Qs.

koute commented:
> Maybe we should add a Clippy config file? Could be the one from Substrate in `.cargo/config.toml`.

Hmmm.... I think we should be fine with the default one here, as this is orders of magnitude smaller so we don't need to disable a bunch of lints like in Substrate.

ggwpez commented:
> Hmmm.... I think we should be fine with the default one here, as this is orders of magnitude smaller so we don't need to disable a bunch of lints like in Substrate.

Its not just disabling, it is also denying some. But only `correctness` and `complexity` as it seems.  
Up to you :shrug:

bkchr commented on src/encode_append.rs:
```diff
@@ -107,46 +97,39 @@ where
+		} else {
+			// We can't update the length as the new length prefix will take up more
+			// space when encoded, so we need to move our data to make space for it.
+			let mut new_vec = Vec::with_capacity(vec.len() * 2);
```
Here we should use `checked_mul`

davxy commented on CHANGELOG.md:
```diff
@@ -5,6 +5,12 @@ All notable changes to this crate are documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this crate adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [3.6.1]
```
Should we add the date?

koute replied:
Yes, I'll do that before I'll merge it (so that the date will actually correspond to the day it'll be released :P)

koute commented:
Added a fix + test for https://github.com/paritytech/parity-scale-codec/issues/456 and also fixed clippy lints in the derive crate (this was as far as I can see necessary to test the fix for #456). Should be good to go once the CI passes.

ggwpez commented on .gitlab-ci.yml:
```diff
@@ -48,6 +48,8 @@ check-clippy:
   script:
     - rustup component add clippy --toolchain stable-x86_64-unknown-linux-gnu
     - time cargo +stable clippy --locked -- -Dwarnings
+    - time cargo +stable clippy --locked -p parity-scale-codec-derive -- -Dwarnings
```
Can you not use something like `--workspace --tests`?

koute replied:
That would run clippy on all of the integration tests, which at this point are not clippy clean, and at this point I didn't feel like also cleaning up all of the tests.

koute commented on tests/clippy.rs:
```diff
@@ -0,0 +1,26 @@
+#[repr(u8)]
+#[derive(Decode, Encode)]
+pub enum CLike {
+	Foo = 0,
```
Yes, good idea. I'll create an issue.
